packages/ast-spec/src/declaration/TSDeclareFunction/spec.ts

```diff
@@ -1,11 +1,35 @@
 import type { AST_NODE_TYPES } from '../../ast-node-types';
 import type { FunctionBase } from '../../base/FunctionBase';
-import type { BlockStatement } from '../../statement/BlockStatement/spec';
 
-// TODO(#1852) - async + declare are semantically invalid together
-export interface TSDeclareFunction extends FunctionBase {
+interface TSDeclareFunctionBase extends FunctionBase {
   type: AST_NODE_TYPES.TSDeclareFunction;
-  body: BlockStatement | undefined;
+  /**
+   * TS1183: An implementation cannot be declared in ambient contexts.
+   */
+  body: undefined;
+  /**
+   * Whether the declaration has `declare` modifier.
+   */
   declare: boolean;
   expression: false;
 }
+
+export interface TSDeclareFunctionWithDeclare extends TSDeclareFunctionBase {
+  /**
+   * TS1040: 'async' modifier cannot be used in an ambient context.
+   */
+  async: false;
+  declare: true;
+  /**
+   * TS1221: Generators are not allowed in an ambient context.
+   */
+  generator: false;
```

Comment:
Tightened

Reply:
Ditto above - the errors are semantic and TS still produces an AST - AND it includes these flags correctly.
So we can't refine these.

Reply:
Aren't we supposed to throw on semantic errors? Wasn't fisker fixing some of these before?